Updated the GlobalIndexUidAggregator to truncate timestamps on keys

ivakegg opened the pull request:
down to the beginning of the day before grouping values with the same key.

ivakegg committed:
Updated the GlobalIndexUidAggregator to truncate timestamps on keys
down to the beginning of the day before grouping values with the same
key.

hgklohr committed:
Merge branch 'integration' into bugfix/globalIndexTimestamps

jschmidt10 commented on warehouse/ingest-core/src/main/java/datawave/ingest/table/aggregator/TruncatingTimestampIterator.java:
```diff
@@ -0,0 +1,118 @@
+     * @return beginning of the day
+     */
+    public static long beginningOfDay(long timestamp) {
+        Calendar c = calendar.get();
```
We could use java8 date/time to do this truncation which would eliminate the need for the ThreadLocal and I believe is a performance win from some quick searching online.

```
Instant truncated = Instant.ofEpochMilli(timestamp).truncatedTo(ChronoUnit.DAYS);
return truncated.toEpochMilli();
```

ivakegg replied:
done

ivakegg committed:
Changed to use java 8 date/time library instead of Calendar

ivakegg committed:
Merge branch 'bugfix/globalIndexTimestamps' of github.com:NationalSecurityAgency/datawave into bugfix/globalIndexTimestamps

ivakegg committed:
Truncate timestamps on keys being returned

ivakegg committed:
Merge branch 'integration' into bugfix/globalIndexTimestamps

apmoriarty committed:
Add a series of integration tests for the global shard index that used
mini accumulo cluster as the backend.

For a number of test cases capture the initial state and the state after
a full major compaction.

ivakegg committed:
Added an iterator that maps uid counts from 0 to 1

ivakegg committed:
Merge branch 'bugfix/globalIndexTimestamps' of github.com:NationalSecurityAgency/datawave into bugfix/globalIndexTimestamps

FineAndDandy committed:
Demonstrate count converstion from Aggregator to RangeStream

hgklohr committed:
Merge branch 'integration' into bugfix/globalIndexTimestamps

ivakegg committed:
Updated test cases with appropriate assertions

ivakegg committed:
Merge branch 'bugfix/globalIndexTimestamps' of github.com:NationalSecurityAgency/datawave into bugfix/globalIndexTimestamps

apmoriarty commented on warehouse/ingest-core/src/test/java/datawave/ingest/table/aggregator/AggregatorIntegrationTests.java:
```diff
@@ -121,6 +136,12 @@ private static void configureTables() throws Exception {
+
+        for (Map.Entry<String,String> entry : client.tableOperations().getConfiguration(UID_AGGREGATOR).entrySet()) {
+            if (entry.getKey().startsWith("table.iterator")) {
+                System.out.println(entry);
```
Could be a log statement

ivakegg replied:
doh., meant to take that out.

ivakegg replied:
removed

apmoriarty commented on warehouse/ingest-core/src/test/java/datawave/ingest/table/aggregator/AggregatorIntegrationTests.java:
```diff
@@ -266,19 +287,19 @@ public void testAddDuplicateUidsWithIncreasingTimeStamps(){
     @Test
-    public void testRemoveThenAddSameTimeStamp(){
+    public void testRemoveThenAddSameTimeStamp() {
         String row = "test_case_eight";
```
I forgot to make this 'getRandomRow()'

ivakegg replied:
done

ivakegg commented on warehouse/ingest-core/src/main/java/datawave/ingest/table/aggregator/GlobalIndexUidAggregator.java:
```diff
@@ -74,15 +73,26 @@ protected boolean isSeenIgnore() {
+     * @return The top key with a truncated timestamp.
+     */
+    @Override
+    public Key getTopKey() {
```
I don't believe this actually makes a difference as the PropogatingIterator is using the key from the underlying source.

ivakegg commented on warehouse/ingest-core/src/main/java/datawave/iterators/TotalAggregatingIterator.java:
```diff
@@ -1,13 +1,24 @@
+
+    @Override
+    public Key getTopKey() {
+        return TruncatingTimestampIterator.getTruncatedTimestampKey(super.getTopKey());
```
This will actually do the job to return the correct timestamp.

ivakegg committed:
review updates

FineAndDandy commented on warehouse/query-core/src/test/java/datawave/query/util/WiseGuysIngest.java:
```diff
@@ -1098,7 +1098,7 @@ private static Value getValueForBuilderFor(String... in) {
     private static Value getValueForNuthinAndYourHitsForFree() {
         Uid.List.Builder builder = Uid.List.newBuilder();
-        builder.setCOUNT(50); // better not be zero!!!!
+        builder.setCOUNT(0); // better not be zero!!!!
```
can you update the comment here to indicate why we are reducing to zero?

ivakegg replied:
sure

apmoriarty commented on warehouse/ingest-core/src/main/java/datawave/iterators/ZeroUIDCountMappingIterator.java:
```diff
@@ -0,0 +1,40 @@
+
+import datawave.ingest.protobuf.Uid;
+
+public class ZeroUIDCountMappingIterator extends WrappingIterator implements SortedKeyValueIterator<Key,Value> {
```
This class needs to be removed now that the zero count case is handled elsewhere

ivakegg replied:
done

ivakegg committed:
review comments
